src/java.base/windows/native/libjava/ProcessHandleImpl_win.c

```diff
@@ -109,27 +109,27 @@ Java_java_lang_ProcessHandleImpl_waitForProcessExit0(JNIEnv* env,
     if (handle == NULL) {
         return exitValue;          // No process with that pid is alive
     }
-    do {
-        if (!GetExitCodeProcess(handle, &exitValue)) {
+    if (!GetExitCodeProcess(handle, &exitValue)) {
+        JNU_ThrowByNameWithLastError(env,
+            "java/lang/RuntimeException", "GetExitCodeProcess");
+    } else if (exitValue == STILL_ACTIVE) {
+        HANDLE events[2];
+        events[0] = handle;
+        events[1] = JVM_GetThreadInterruptEvent();
+        // Interrupt and process termination are treated equally.
+        // In case of an interrupt, if the process is still active,
+        // the method returns STILL_ACTIVE (259).
+        if (WaitForMultipleObjects(sizeof(events)/sizeof(events[0]), events,
+                                   FALSE,    /* Wait for ANY event */
+                                   INFINITE) /* Wait forever */
+                == WAIT_FAILED) {
+            JNU_ThrowByNameWithLastError(env,
+                "java/lang/RuntimeException", "WaitForMultipleObjects");
+        } else if (!GetExitCodeProcess(handle, &exitValue)) {
```

Comment:
This can return STILL_ACTIVE if there is a spurious thread interrupt.
The interrupt is presumably present to keep the thread from being stuck in a blocking windows os call.

The calling code in ProcessHandleImpl is agnostic to platform and would report the process had exited.

Can the risk of incorrectly reporting the process exit be mitigated?

If the Thread is legitimately been interrupted, Thread.interrupted() would be true and the reaper could exit.
If false, it could retry the waitForProcessExit().

Reply:
I only left it here because the wait for interrupt event was already present. Is being stuck in a blocking os call a bad thing? If not, I can drop the interrupt event, and wait on the process handle only.

Reply:
I think waiting on JVM_GetThreadInterruptEvent() is necessary during VM shutdown to allow the blocked thread to exit cleanly.

Reply:
Interesting. Could you point me to a related bug / documentation / evidence? I tested the WaitFocSingleObject version (commit 6524747a5ebc51c760b14e90309c5f18b58b20e2) on Windows 11, and I didn't observe any problems with shutting down the VM

Reply:
I don't have a test or counter example and don't have a detailed understanding of the blocking needs of Hotspot.

Reply:
Another reason to retain the checking of GetThreadInterruptEvent is to be belt and suspenders against the Java code changing and opening up a potential error. At the moment, the reaper thread is encapsulated and not likely to get an interrupt, but that might not always be the case and someone changing the Java code might be unaware of the native code details.  $.02

Reply:
I did some research, and now I'm pretty confident that WaitForSingleObject is the right option:
- the thread involved here is a daemon thread, i.e. it doesn't prevent JVM exit
- the Unix version of `waitForProcessExit0` ignores interrupts and signals, it only ends when the wait succeeds or fails
- I didn't observe any problems with exiting the JVM while waiting on a process handle
- it's not clear to me what action we would take if the thread was cancelled; completing a CompletableFuture with an InterruptedException doesn't feel right.

Reply:
ok, I'll buy that argument